driver-core/src/test/functional/com/mongodb/client/model/expressions/ControlExpressionsFunctionalTest.java

```diff
@@ -0,0 +1,205 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.model.expressions;
+
+import org.bson.Document;
+import org.junit.jupiter.api.Test;
+
+import java.time.Instant;
+import java.util.function.Function;
+
+import static com.mongodb.client.model.expressions.Expressions.of;
+import static com.mongodb.client.model.expressions.Expressions.ofIntegerArray;
+import static com.mongodb.client.model.expressions.Expressions.ofMap;
+import static com.mongodb.client.model.expressions.Expressions.ofNull;
+
+class ControlExpressionsFunctionalTest extends AbstractExpressionsFunctionalTest {
+
+    @Test
+    public void applyTest() {
+        Function<IntegerExpression, IntegerExpression> decrement = (e) -> e.subtract(of(1));
+        // "nested functional" function application:
+        assertExpression(
+                2 - 1,
+                decrement.apply(of(2)),
+                "{'$subtract': [2, 1]}");
+        // "chained" function application:
+        assertExpression(
+                2 - 1, // = 0
+                of(2).apply(decrement),
+                "{'$subtract': [2, 1]}");
+        // the parameters are reversed, compared to Java's function.apply
+    }
+
+    @Test
+    public void switchTest() {
+        // https://www.mongodb.com/docs/manual/reference/operator/aggregation/switch/
+        assertExpression("a", of(0).switchMap(on -> on.is(v -> v.eq(of(0)), v -> of("a"))));
+        assertExpression("a", of(0).switchMap(on -> on.isNumber(v -> of("a"))));
+        assertExpression("a", of(0).switchMap(on -> on.lte(of(9), v -> of("a"))));
+
+        // test branches
+        Function<IntegerExpression, BooleanExpression> isOver10 = v -> v.subtract(10).gt(of(0));
+        Function<IntegerExpression, StringExpression> s = e -> e
+                .switchMap((Branches on) -> on
+                        .eq(of(0), v -> of("A"))
+                        .lt(of(10), v -> of("B"))
+                        .is(isOver10, v -> of("C"))
+                        .defaults(v -> of("D")))
+                .toLower();
+
+        assertExpression("a", of(0).apply(s));
+        assertExpression("b", of(9).apply(s));
+        assertExpression("b", of(-9).apply(s));
+        assertExpression("c", of(11).apply(s));
+        assertExpression("d", of(10).apply(s));
+    }
+
+    @Test
+    public void switchTypesTest() {
+        Function<IntegerExpression, StringExpression> label = expr -> expr.switchMap(on -> on
+                .isBoolean(v -> v.asString().concat(of(" - bool")))
+                .isNumber(v -> v.asString().concat(of(" - number")))
+                .isString(v -> v.asString().concat(of(" - string")))
+                .isDate(v -> v.asString().concat(of(" - date")))
+                .isArray((ArrayExpression<IntegerExpression> v) -> v.sum(a -> a).asString().concat(of(" - array")))
```

Comment:
Why does this require specifying the type of `v`?

Reply:
Alternatively:
`.<IntegerExpression>isArray(v -> ...`

`isArray` checks that the value is an array, but cannot check the type of the array (I see no reasonable way to do this). So the method is responsible for checking whether it's an array, but the user is responsible for "asserting" the type of the array (much like they assert the type during doc.getInteger).